### PR TITLE
Issue 13 - Storing token in database

### PR DESCRIPTION
**Pull request by pfmluis**

### CODE REVIEW

File: app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java [hunk 1]
```diff
@@ -226,4 +229,13 @@ public void goToActivityAndFinish(Class activityToGo) {
         IntentNavigationUtils.goToActivityAndFinish(LoginActivity.this, activityToGo,
                 Collections.emptyMap());
     }
+
+    private List<String> initializerUserTokens(){
+        List<String> tokens = new ArrayList<>();
```
Comment (RicardooCorreia) [inline app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java]: Are you sure that this is the way that you get the FCM token?

Comment (RicardooCorreia) [reply app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java]: https://firebase.google.com/docs/cloud-messaging/android/client#sample-register

File: app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java [hunk 2]
```diff
@@ -13,8 +26,44 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
-        //TODO HANDLE TOKEN CHANGES HERE
+    public void onNewToken(String token) {
+        Log.d("REFRESH TOKEN", token);
+        FirebaseAuth mAuth = FirebaseAuth.getInstance();
+        if(mAuth.getCurrentUser() != null){
+            UserRepository.getInstance().getByField("key", FirebaseAuth.getInstance()
+                    .getCurrentUser().getUid()).addValueEventListener(new ValueEventListener() {
+                @Override
+                public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
+                    saveTokenOnRealTimeDB(dataSnapshot, token);
```
Comment (RicardooCorreia) [inline app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java]: Add a level of abstraction here. At this code level, we don't need to know that the token is saved on the "Realtime DB", we just need to know that it's saved.

File: app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java [hunk 3]
```diff
@@ -13,8 +26,44 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
+
+
     @Override
-    public void onNewToken(String s) {
-        //TODO HANDLE TOKEN CHANGES HERE
+    public void onNewToken(String token) {
+        Log.d("REFRESH TOKEN", token);
+        FirebaseAuth mAuth = FirebaseAuth.getInstance();
+        if(mAuth.getCurrentUser() != null){
+            UserRepository.getInstance().getByField("key", FirebaseAuth.getInstance()
```
Comment (RicardooCorreia) [inline app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java]: Why this way?
When you have the current user you just need to get that user, go to child "tokens" and add the new to token to it

File: app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java [hunk 4]
```diff
@@ -13,8 +26,44 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
+                    saveTokenOnRealTimeDB(dataSnapshot, token);
+                }
+
+                @Override
+                public void onCancelled(@NonNull DatabaseError databaseError) {
+                    Log.e("", "Failed to load specified user");
+                }
+            });
+        } else {
+            PreferenceManager preferenceManager = PreferenceManager.getInstance(getApplicationContext());
```
Comment (RicardooCorreia) [inline app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java]: Same as above, I don't get why you need PreferenceManager

File: app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java [hunk 5]
```diff
@@ -153,6 +162,13 @@ private void firebaseSignIn(String email, String password) {
                                     .ifPresent(firebaseUser ->
                                             firebaseUser.updateProfile(profileUpdates));
 
+                            getCurrentDeviceToken(token -> {
+                                List<String> tokens = new ArrayList<>();
```
Comment (RicardooCorreia) [inline app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java:199]: You can use: `Collections.singletonList(token)`

File: app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java [hunk 6]
```diff
@@ -163,6 +179,27 @@ private void firebaseSignIn(String email, String password) {
                 });
     }
 
+    private void saveCurrentUser(String displayName, List<String> tokens) {
+        FirebaseUser currentUser = mAuth.getCurrentUser();
+        UserRepository.getInstance().save(
+                currentUser.getUid(),
+                User.Builder.aUser()
```
Comment (RicardooCorreia) [inline app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java:219]: Ctrl + Alt  + L

Is triggering me a little

File: app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java [hunk 7]
```diff
@@ -163,6 +179,27 @@ private void firebaseSignIn(String email, String password) {
...
+                .withId(currentUser.getUid())
+                .withName(displayName)
+                .withTokens(tokens).build()
+        );
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.N)
+    private void getCurrentDeviceToken(Consumer<String> consumer){
+        FirebaseInstanceId.getInstance().getInstanceId().addOnSuccessListener(new OnSuccessListener<InstanceIdResult>() {
```
Comment (RicardooCorreia) [inline app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java]: You can replace this by lambda function:
`FirebaseInstanceId.getInstance().getInstanceId().addOnSuccessListener(instanceIdResult -> consumer.accept(instanceIdResult.getToken()));`

File: app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java [hunk 8]
```diff
@@ -13,8 +31,38 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
+                public void onCancelled(@NonNull DatabaseError databaseError) {
+                    Log.e("GET USER ERROR", databaseError.toException().toString());
+                }
+            });
+        }
     }
+
+    @RequiresApi(api = Build.VERSION_CODES.N)
+    private List<String> getUserTokens(DataSnapshot snapshot) {
+        return Objects.isNull(snapshot) ? new ArrayList<String>() : (List<String>) snapshot.getValue();
```
Comment (RicardooCorreia) [inline app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java]: You don't need to specify <String> in ArrayList.

File: app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java [hunk 9]
```diff
@@ -13,8 +31,38 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
+                @RequiresApi(api = Build.VERSION_CODES.N)
+                @Override
+                public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
+                    List<String> tokens = getUserTokens(dataSnapshot);
+
+                    if (!tokens.contains(token)) {
+                        tokens.add(token);
+                    }
+
+                    userTokensReference.setValue(tokens);
```
Comment (RicardooCorreia) [inline app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java:52]: Have you tested login in with the same account in multiple devices? To see if all the tokens are saved?

File: app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java (hunk 5 above)
Comment (RicardooCorreia) [reply app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java:199]: Saving a lot of unnecessary code:
`saveCurrentUser(displayName, Collections.singletonList(token));`

**Review by RicardooCorreia — CHANGES_REQUESTED** [review]
After these minor changes to clean the code (and tested in multiple devices), this branch can be merged to master.

File: app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java (hunk 6 above)
Comment (pfmluis) [reply app/src/main/java/com/estg/masters/pedwm/smarthome/activity/LoginActivity.java:219]: Cry me a river

File: app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java (hunk 9 above)
Comment (pfmluis) [reply app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java:52]: Gonna try it now

File: app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java (hunk 8 above)
Comment (pfmluis) [reply app/src/main/java/com/estg/masters/pedwm/smarthome/services/MyFirebaseMessagingService.java]: Ooopsie, my bad

**Pull request merged**